sockeye/translate.py

```diff
@@ -67,60 +69,75 @@ def main():
                                                   *sockeye.inference.load_models(context,
                                                                                  args.max_input_len,
                                                                                  args.beam_size,
+                                                                                 args.batch_size,
                                                                                  args.models,
                                                                                  args.checkpoints,
                                                                                  args.softmax_temperature,
                                                                                  args.max_output_length_num_stds))
-        read_and_translate(translator, output_handler, args.input)
+
+        logger.info("Using batches of size %d", args.batch_size)
+        read_and_translate(translator, output_handler, args.chunk_size, args.input)
 
 
 def read_and_translate(translator: sockeye.inference.Translator, output_handler: sockeye.output_handler.OutputHandler,
-                       source: Optional[str] = None) -> None:
+                       chunk_size: int, source: Optional[str] = None) -> None:
     """
     Reads from either a file or stdin and translates each line, calling the output_handler with the result.
 
     :param output_handler: Handler that will write output to a stream.
     :param translator: Translator that will translate each line of input.
+    :param chunk_size: The size of the portion to read at a time from the input.
     :param source: Path to file which will be translated line-by-line if included, if none use stdin.
     """
+    def grouper(iterable: Iterable, size: int) -> Iterable:
+        """
+        Collect data into fixed-length chunks or blocks
+        without either dicarding underfilled chunks or padding them
+        """
+        it = iter(iterable)
+        while True:
+            chunk = list(itertools.islice(it, size))
+            if not chunk:
+                return
+            yield chunk
 
     source_data = sys.stdin if source is None else sockeye.data_io.smart_open(source)
 
     logger.info("Translating...")
 
-    i, total_time = translate_lines(output_handler, source_data, translator)
+    total_time, total_lines = 0.0, 0
+    for chunk in grouper(source_data, chunk_size):
+        lines, chunk_time = translate(output_handler, chunk, translator)
+        total_lines += lines
```

Comment:
couldn't you just do `total_lines += len(chunk)`? And otherwise I'd rename to `num_lines`.

Reply:
True. I think we can get rid of accumulating total_lines, since it will be always equal to len(source_data)

Reply:
That  I am not sure about since source_data is a generator, right? Len would the consume it to calculate length which is bad :)

Reply:
it's actually a list (see line 99)

Reply:
True, nevermind my comment then.